### DISCUSSION TITLE
Jitter in RT communication under system load (Linux)

### DISCUSSION THREAD

**Discussion by Schrolli91** [eclipse-iceoryx/iceoryx#1374] [Q&A]
Hey folks, 

we are currently investigating the suitability of Iceoryx for applications with hard realtime requirements under Linux.  We noticed a lot of unexpected jitter, especially compared to a custom/self-made lib we are using internally which uses a shared memory mechanism. After some tests we are not sure where this extreme jitter comes from. It occurs mainly under load of the system.  

The only difference we see to our simpler tool is the usage of atomic operations (CAS) in lockfree-queue handling and chunk management. According to the "perf" tool we saw that these operations have the greatest overhead and therefore the highest impact on latency during Pub/Sub communication. 

Maybe one of you has an idea what the reason could be why Iceoryx is so much more susceptible to jitter, or what we should investigate further. 

Best regards

---

**Operating system:** 
Debian 11 - Kernel 5.10.84-rt58 
Iceoryx Version 2.0.1 
Hardware: Industrial PC with Quadcore CPU 

**Compiler version:** 
gcc 10.2.1 

**Observed result or behavior:** 
We see a lot of jitter within the communication. This occurs especially under increased system load. 
A custom lib based on a simple "shmem" approach (p2p/one-way) does not show this extreme jitter. 

**Expected result or behavior:** 
Deterministic transmission times of the packets by the usage of Iceoryx. 

**Conditions where it occurred / Performed steps:** 
*Test Application:* 
Tests were executed with a simple Publisher/Subscriber application based on provided examples. 
Both Application subscribe to each other and proceed to ping-pong messages. 
One million packets were sent for warm-up, and then RTTs were measured individually for 10 million messages.  

*System configuration:*  
- RouDi 
Affinity: `taskset –c 2 chrt –f 80`  
Prio: `FIFO 80` 
Pinned to Core 2  

- Publisher 
Affinity: `taskset –c 2 chrt –f 80`  
Prio: `FIFO 80` 
Pinned to Core 2 

- Subscriber 
Affinity: `taskset –c 3 chrt –f 80` 
Prio: `FIFO 80` 
Pinned to Core 3 

Stress/Systemload is generated via: 
`stress-ng --cpu 4 --io 2 --vm 2 --vm-bytes 128M --fork 4 --timeout 0` 

---

Here is a comparison of the jitter we could observe using Iceoryx compared to our simpler internal library.
(The absolute values are not the problem, but the enormous jitter at Iceoryx). 

**Iceoryx (untyped API) – polling – 80 bytes – with stress** 
![iceoryx](https://user-images.githubusercontent.com/6043382/170014487-32c6cdcb-bfdc-4873-920e-ea680be005b8.png)

**Custom implementation – polling – 80 bytes – with stress** 
![custom_impl](https://user-images.githubusercontent.com/6043382/170014589-c3ee003a-e0bb-4db3-98a3-44bcb5331a9a.png)

**Comment by elBoberido** [comment]
@Schrolli91 thanks for doing this tests. Could you please repeat them with `RouDi` on a dedicated core? We suspect that this is causing the jitter since RouDi has some threads which run every 100ms and will compete for CPU time with the publisher.

Another question, does your custom implementation also have a broker? If yes, does it run on the same core as one of the clients?

**Reply by Schrolli91** [reply]
Hey @elBoberido, thanks for your fast reply. 

Our custom implementation does not have a broker. We use a shared memory for a one-way transfer. 
But our implementation is not a true zero copy, we use one single `memcpy`. 

--- 

We have also tried to isolate the RouDi. But our measurement does not show any improvement in the jitter values.  

*System configuration:*  
Core 0 and Core 1 are used for all the main linux stuff.

RouDi 
Affinity: `taskset –c 1 chrt –f 80`  
Prio: `FIFO 80` 
Pinned to Core 1 (is not a RT isolated Core)  

Publisher 
Affinity: `taskset –c 2 chrt –f 80`  
Prio: `FIFO 80` 
Pinned to Core 2 

Subscriber 
Affinity: `taskset –c 3 chrt –f 80` 
Prio: `FIFO 80` 
Pinned to Core 3 

---

What we have seen with the `perf record` tool (https://man7.org/linux/man-pages/man1/perf-record.1.html) is that the "queue pop" and "decrement ref counter" functions are very expensive in terms of processor time. Do you think thats ok?
![perf_dump](https://user-images.githubusercontent.com/6043382/170258323-eaad334b-fe54-47c8-9941-395cb381dd4b.png)

Are there any other ideas what we can try against our problem? Or do you think this is a normal problem under heavy load of the system with iceoryx?

Best regards,
Basti

**Reply by elfenpiff** [reply]
@Schrolli91 

We use a different queue when we have one publisher to many subscriber. The default that we support are multiple publishers which publish on the same topic to many subscribers. Maybe the cmake switch `-DONE_TO_MANY_ONLY=ON` solves the issue when you can restrict your setup to one publisher per topic. Here I would expect that the cpu load in `tryGetChunk` is maybe reduced.

I think the cpu load in `{increment|decrement}ReferenceCounter` is maybe caused by cache misses. When you look into the application you see a `nullptr` check and then a dereferencing of `m_chunkManagement` to acquire `m_referenceCounter` (here cache misses may occur since the whole ChunkManagement object is loaded into the cache just to access a member). It may improve the performance when `m_referenceCounter` is stored as additional member variable pointer in `SharedChunk`.
I will try to add this as a playground branch somewhere and ping you so that you can try it out.

How do you sync your memory during the one way transfer? Do you use a mutex?

**Reply by Schrolli91** [reply]
hey @elfenpiff
we will try the compile flag `-DONE_TO_MANY_ONLY=ON` and give feedback tomorrow. Thanks for the tip first.
We will also do a `perf-record` recording again.

> I will try to add this as a playground branch somewhere and ping you so that you can try it out.

That would be great. Thanks a lot for that.

> How do you sync your memory during the one way transfer? Do you use a mutex?

One of our test applications does sync through a sempahore, but I'm not sure if this is the case with the application for our load test.
I am out of the office right now. But will check again tomorrow and let you know.

**Reply by Schrolli91** [reply]
@elfenpiff 
We made a new `perf-record` recording with the `-DONE_TO_MANY_ONLY=ON` flag you mentioned. Here you can see the result of our recording:

![perf_dump_new](https://user-images.githubusercontent.com/6043382/170669567-a46e5d2c-85d0-4aab-b531-5e2d16152fb4.png)

Unfortunately the jitter in our measurements is not appreciably better than before but more constant overall. All in all, a slight improvement, but not sufficient for our purposes.

---

> How do you sync your memory during the one way transfer? Do you use a mutex?

Yes, memory access is locked via mutex.

**Comment by elfenpiff** [comment]
@Schrolli91 I played around with it over the long weekend and at the moment I suspect maybe our lock-free algorithms. They have the advantage that we do not require a mutex which is perfect for a safety-critical system but may have the disadvantage that they require longer under high CPU load. In most lock-free structures you find some kind of loop where the work is performed and if the structure is unchanged after the work is done you change the structure but if the structure was modified you start again.
And when I see the `IndexQueue::pop` I suspect that this is exactly what happens here.

If this is the case we could implement a queue which is guarded by a mutex which does not have this problem at all. This maybe solves your problem but would not be usable in a safety critical context since blocking calls are forbidden.

**Reply by elfenpiff** [reply]
@Schrolli91 

Your playground branch is in https://github.com/apexai/iceoryx named `iox-#1374-mutex-based-queue` .

If my suspicions with the lock-free issues are correct this branch should reduce the jittering and increase the performance of your setup.

If this works for you, we can create an issue and introduce this with tests and all the stuff into iceoryx, when it's still jittering we have to dig deeper.

**Reply by Schrolli91** [reply]
Hey, thanks for providing the branch.

We have recorded our measurements again the last days.
Unfortunately, the jitter is not better, on the contrary, it has become worse.
See also the following screenshot:

![jitter_new](https://user-images.githubusercontent.com/6043382/171593409-183a550a-4319-485c-abc4-66f79039a13c.png)

In the upper area it is almost identical. In the lower range it has become a bit more due to the blocking mutex.

In the meantime we are not so sure if the actual jitter really has something to do with iceoryx or if it comes from the OS itself (maybe linux housekeeping?).
We especially noticed the regular dropouts, about every ~3 million packets (about 5 - 8 seconds).

**Reply by elfenpiff** [reply]
@Schrolli91 At least we know now that our lock free queue is pretty fast :D . 

The jitter could come from the OS itself. Maybe it originates in the two additional threads that roudi brings. In another problem they were also suspected to cause some issues and I created a playground branch to unify them into one thread. Could you checkout [iox-#1361-mitigate-process-keepalive-timeout](https://github.com/ApexAI/iceoryx/tree/iox-%231361-mitigate-process-keepalive-timeout) and benchmark it one last time.

The background threads are firing every 50 or 10 ms and handle the keepalive messages from the apps as well as the service discovery. And they come with an expensive context switch which may caused the jittering. Or is your benchmark somehow publicly available so that I could take a look myself?

**Reply by elBoberido** [reply]
@elfenpiff @Schrolli91 there is also a background thread in each application which sends a keep alive via a unix domain socket to roudi. This might also be a cause for the jittering.

We could try two things:
- disable the transmission via the unix domain socket
- disable the whole thread

The context switch between the treads should not be too expensive but maybe the code path to the unix domain socket has a significant overhead and we might notice this if the thread is executed on the same core.

**Reply by Schrolli91** [reply]
hey @elfenpiff

unfortunately your suggestion on the branch `iox-#1361-mitigate-process-keepalive-timeout` did not bring the desired success either.
> **FYI:**
> I have seen that this branch belongs to the following PR #1370.
> We have noticed that on prio `FIFO80` under high CPU load an error occurs.
> This error occurs in the original as well as on the PR Branch!
> ![roudi_error](https://user-images.githubusercontent.com/6043382/171814298-23137d42-5cd4-4d85-a639-abc67af863cd.png)

We have made a record again (Prio: `RR30`) - but all in all a little bit worse than before (without any modiification):
![jitter_new_1](https://user-images.githubusercontent.com/6043382/171809059-4bda245d-8108-4431-9908-5deb08e4a9e8.png)

@elfenpiff, @elBoberido
As you can see in the following graph, we regularly have significant latency peaks. These occur in an interval of about 10 seconds. Maybe this time interval looks familiar to you?
![jitter_interval](https://user-images.githubusercontent.com/6043382/171811232-1519d9e3-937e-41d5-bbed-5d5734fbc4a3.png)

@elBoberido 
What do you think would be the easiest way to turn off the thread with the keep alive messages through the unix domain socket?

**Reply by elfenpiff** [reply]
@Schrolli91 @elBoberido 

I took a look at all the background threads which could possibly run and deactivated them. So on the application side the keep alive message thread is gone and on the roudi side I deactivated the mempool, port and process introspection threads. So the overhead of threads and the unix domain socket keepalive message should be gone.

When you take a look at roudi with `ps -T -p ${ROUDI_PID}` you should see:
```
# ps -T -p 34822
    PID    SPID TTY          TIME CMD
  34822   34822 pts/2    00:00:00 iox-roudi
  34822   34823 pts/2    00:00:00 IPC-msg-process
```

In any other iceoryx application only the main thread should be active:
```
# ps -T -p 34875
    PID    SPID TTY          TIME CMD
  34875   34875 pts/3    00:00:00 iox-cpp-publish
```

I pushed everything on the `iox-#1361-mitigate-process-keepalive-timeout` branch maybe we get better results now.

**Reply by Schrolli91** [reply]
@elfenpiff Now we get an `SEG_FAULT` immediately on this Branch when subscriber connects.
![roudi_seg_fault](https://user-images.githubusercontent.com/6043382/171834320-304c1807-bc15-4c9b-9d35-24dff3759b00.png)

**Reply by elfenpiff** [reply]
@Schrolli91 Sorry, my bad. In my wild introspection removal action I forgot to cleanly remove the process introspection stuff. 

Now it is working!

**Reply by elBoberido** [reply]
@Schrolli91 @elfenpiff I think I have now an explanation for the pattern. I think it's an interference between the two keep-alive threads which results in some kind of [beat](https://en.wikipedia.org/wiki/Beat_(acoustics)).

I count around 50 high latency measurements in the first period of the peaks and it looks like the following periods have a similar count. Assuming that now and then the latency is not that much affected by the keep-alive thread it is quite close to 67 measurements or one measurement each 150 ms. With two applications having two background threads which are triggered each 300ms, this means on average every 150 ms one of the applications is affected by this thread. The peaks each 10 seconds are the result of the keep-alive thread being scheduled at the same time on both applications. Since there is some jittering on the thread scheduling, especially under high load, this is quite likely.

The question is now if it is the thread alone or is it something which is done in the tread and therefore we would need to have a run without sending the keep-alive via the unix domain socket.

**Reply by Schrolli91** [reply]
New run on `iox-#1361-mitigate-process-keepalive-timeout` - much better.
And no error on "stressed" `FIFO80` with `-m off` - Thanks for the tipp.
![new_jitter](https://user-images.githubusercontent.com/6043382/171845466-2d514b40-d17e-488a-83d1-4d0c8a0b136f.png)

@elfenpiff 
In this branch, is the keep alive over the Unix domain sockets completely disabled as was desired by @elBoberido ? Then we would now have the corresponding results here.

**Reply by elfenpiff** [reply]
@Schrolli91 @elBoberido The keep alive over unix domain sockets is completely disabled on this branch.

**Reply by elBoberido** [reply]
@elfenpiff just to be sure, is is only the keep alive over unix domain sockets or is the whole thread in the application disabled. If it is the latter, it would be nice to have an additional measurement with the thread running and just the transmission disabled. The reason for this is that this thread currently also does some cleanup on shutdown in case a blocking publisher/subscriber is used.

**Reply by elfenpiff** [reply]
@Schrolli91 @elBoberido I created an issue where we add the feature of disabling the runtime keepalive message, see #1380

**Reply by elfenpiff** [reply]
@elBoberido the whole thread in the application is disabled.

I rebased a little and pushed a variant where all the threads are running but the keepalive message is not send and the monitoring in roudi is always off.

@Schrolli91 could you measure it again and tell us if the jittering is still gone?

**Reply by Schrolli91** [reply]
@elfenpiff, @elBoberido 
Fresh pulled on same branch - Jitter is still gone with this variant.

By the way, We would like to thank you for your efforts so far.

**Reply by elfenpiff** [reply]
@Schrolli91 no problem at all. You revealed a real issue so we have to thank you as well for debugging and benchmarking this together with us.

If you like you can join our developer meetup next thursday, see: https://github.com/eclipse-iceoryx/iceoryx/wiki/Developer-meetup then we will meet virtually "in person".

**Reply by elBoberido** [reply]
@Schrolli91 thanks for your input. Btw. how did you create those charts?

**Reply by Schrolli91** [reply]
@elBoberido we created these charts using matplotlib/numpy in Python.
Our benchmark application saves the measurements on the drive and we then create these plots using the script.

**Comment by elfenpiff** [comment]
@Schrolli91 

> I have seen that this branch belongs to the following PR https://github.com/eclipse-iceoryx/iceoryx/pull/1370.
> We have noticed that on prio FIFO80 under high CPU load an error occurs.
> This error occurs in the original as well as on the PR Branch!

This problem you can mitigate by starting roudi with `./build/iox-roudi -m off` which turns off the monitoring. Here roudi tracks the incoming keepalive messages and when no message was received for 1500ms roudi kills the app which causes your problem.

**Comment by elfenpiff** [comment]
@Schrolli91 I created the issue #1436 which should solve your jitter problem when you turn of the monitoring in roudi. I would like to add you as a reviewer if it's alright with you so that you are able to test if the jittering is really gone.

**Reply by Schrolli91** [reply]
yes please add me and we will check it once again.
